conftest.py

```diff
@@ -26,7 +26,12 @@
 
 
 def pytest_sessionstart():
-    """ Download sct_testing_data prior to test collection. """
+    """Perform actions that must be done prior to test collection."""
+    # Use a non-interactive backend so that no GUI plots will interrupt the test suite.
+    # (NB: We do this here to ensure it is set before `matplotlib` is first imported.)
+    os.environ["MPLBACKEND"] = 'Agg'
```

Comment:
Note: This change is slightly different than what @kousu suggested in https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3388#issuecomment-861620956.

If we set `MPLBACKEND` in `tests.yml`, then interactive plots will still interrupt the test suite when `pytest` is run locally.

Reply:
Ah, true. :+1: 

Can we do both? If we later handle https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3392 / https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3388 by making sure the non-interactive tests never have to call any non-interactive code we can get rid of this smoothly.

Reply:
Sorry, I'm still a bit unclear still on the benefits of doing both. 

In the future, if we remove interactive plotting code, wouldn't we not need `MPLBACKEND=Agg` in either place?

Reply:
https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3388#issuecomment-846091012

> to test [..] I suggest unconditionally setting the backend to Agg (that is what we do in the Matplotlib tests). It avoids a whole class of pain points (related to making sure you have the event loop running enough, etc) and will make your tests run faster!

Reply:
Right!

In that case, what's the benefit of setting it in the CI config instead of setting it within `pytest`?

Reply:
Well, it'll keep CI faster, for one.

I guess I was thinking that we know the CI machines are headless, but the matplotlib doesn't, and this is how we communicate that fact to it globally.

Communicating the same via pytest feels like more of a workaround. Maybe you want to run the tests locally in your GUI just to see what they do sometimes?

Reply:
> Maybe you want to run the tests locally in your GUI just to see what they do sometimes?

Ah, good point. And if we're setting it in pytest, there's not a great way to unset it beyond going it and modifying that LOC, which is a little problematic.

At the very least, to address this we could do this instead:

```suggestion
    # Use a non-interactive backend so that no GUI plots will interrupt the test suite.
    # (NB: We do this here to ensure it is set before `matplotlib` is first imported.)
    if 'MPLBACKEND' not in os.environ:
        os.environ["MPLBACKEND"] = 'Agg'
```

That way, a dev can specify e.g. `export MPLBACKEND=TkAgg` locally to effectively turn GUI testing back on.